test/Autofac.Test/Features/OpenGenerics/OpenGenericDelegateTests.cs

```diff
@@ -0,0 +1,141 @@
+
+            Assert.Throws<ComponentNotRegisteredException>(() => container.Resolve<IInterfaceB<string>>());
+        }
+
+        [Fact]
+        public void MultipleTypeArgsSupported()
+        {
+            var builder = new ContainerBuilder();
+
+            builder.RegisterGeneric((ctxt, types) => Activator.CreateInstance(typeof(ImplementationMultiType<,>).MakeGenericType(types)))
+                   .As(typeof(IInterfaceMultiType<,>));
+
+            var container = builder.Build();
+
+            var instance = container.Resolve<IInterfaceMultiType<string, int>>();
+
+            var implementedType = instance.GetType().GetGenericTypeDefinition();
+
+            Assert.Equal(typeof(ImplementationMultiType<,>), implementedType);
+        }
+
+        [Fact]
+        public void VariableTypeArgsLengthSupported()
+        {
+            var builder = new ContainerBuilder();
+
+            builder.RegisterGeneric((ctxt, types) =>
+            {
+                var chosenType = types.Length == 2 ? typeof(ImplementationMultiType<,>) : typeof(ImplementationA<>);
+
+                return Activator.CreateInstance(chosenType.MakeGenericType(types));
+            })
+            .As(typeof(IInterfaceMultiType<,>))
+            .As(typeof(IInterfaceA<>));
+
+            var container = builder.Build();
+
+            var instance = container.Resolve<IInterfaceA<string>>();
+            var implementedType = instance.GetType().GetGenericTypeDefinition();
+            Assert.Equal(typeof(ImplementationA<>), implementedType);
+
+            var multiInstance = container.Resolve<IInterfaceMultiType<string, int>>();
+            implementedType = multiInstance.GetType().GetGenericTypeDefinition();
+            Assert.Equal(typeof(ImplementationMultiType<,>), implementedType);
+        }
+
+        [Fact]
+        public void GenericDelegateCanReceiveParameters()
+        {
+            var builder = new ContainerBuilder();
+
+            List<Parameter> passedParameters = null;
+
+            builder.RegisterGeneric((ctxt, types, parameters) =>
+            {
+                passedParameters = parameters.ToList();
+
+                return Activator.CreateInstance(typeof(ImplementationA<>).MakeGenericType(types));
+            })
+            .As(typeof(IInterfaceA<>));
+
+            var container = builder.Build();
+
+            var instance = container.Resolve<IInterfaceA<int>>(new TypedParameter(typeof(bool), true));
+
+            Assert.Collection(
+                passedParameters,
+                p => Assert.IsType<TypedParameter>(p));
+        }
+
+        [Fact]
+        public void CastExceptionIfDelegateReturnsBadObject()
+        {
+            var builder = new ContainerBuilder();
+
+            builder.RegisterGeneric((ctxt, types) => "bad")
+                   .As(typeof(IInterfaceA<>));
+
+            var container = builder.Build();
+
+            Assert.Throws<InvalidCastException>(() => container.Resolve<IInterfaceA<string>>());
```

Comment:
I'm wondering if this should be a `DependencyResolutionException` with the cast exception inside. That way we know what was _trying_ to be resolved, and then the inner exception explains more about why it got jacked up.

Reply:
Also, additional test case - what happens if the delegate throws? Like, maybe it throws `DivideByZeroException`. What bubbles up?

Reply:
I did debate whether to add specific behaviour to check the generated type, but I'm now wondering if instead I should throw a try/catch around [the actual cast](https://github.com/autofac/Autofac/blob/develop/src/Autofac/ResolutionExtensions.cs#L314) and throw the nested error.

Reply:
There are a few of those casts, like [`ResolveOptionalService`](https://github.com/autofac/Autofac/blob/develop/src/Autofac/ResolutionExtensions.cs#L617) has a few. I think it'd be fine to handle it at that level, it'd just have to be handled in all the spots.

Reply:
OK, the cast wrapping is in.  I've also added that additional test case, which is treated like any activator exception, and wrapped as normal.